In [1]:
import tensorflow as tf
import numpy as np
from tensorflow.keras.datasets import imdb

In [2]:
tf.__version__

'2.4.0'

# Pre-Processing

## Database Parameters

In [3]:
number_of_words = 20000
max_len = 100

## Loading IMDB Dataset

In [4]:
(x_train,y_train), (x_test,y_test) = imdb.load_data(num_words=number_of_words)

17465344/17464789 [==============================] - 0s 0us/step


<string>:6: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/datasets/imdb.py:159: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  x_train, y_train = np.array(xs[:idx]), np.array(labels[:idx])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/datasets/imdb.py:160: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 

In [5]:
x_train.shape #25.000 Texts

(25000,)

List (Words represented by numbers) ex: 14 == Hello

Pre Processed database don't need to map words to transform text numbers

![](https://miro.medium.com/max/700/1*ZtGzGoYT7VFf5uU0JFsFSw.png)

In [6]:
x_train

array([list([1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65, 458, 4468, 66, 3941, 4, 173, 36, 256, 5, 25, 100, 43, 838, 112, 50, 670, 2, 9, 35, 480, 284, 5, 150, 4, 172, 112, 167, 2, 336, 385, 39, 4, 172, 4536, 1111, 17, 546, 38, 13, 447, 4, 192, 50, 16, 6, 147, 2025, 19, 14, 22, 4, 1920, 4613, 469, 4, 22, 71, 87, 12, 16, 43, 530, 38, 76, 15, 13, 1247, 4, 22, 17, 515, 17, 12, 16, 626, 18, 19193, 5, 62, 386, 12, 8, 316, 8, 106, 5, 4, 2223, 5244, 16, 480, 66, 3785, 33, 4, 130, 12, 16, 38, 619, 5, 25, 124, 51, 36, 135, 48, 25, 1415, 33, 6, 22, 12, 215, 28, 77, 52, 5, 14, 407, 16, 82, 10311, 8, 4, 107, 117, 5952, 15, 256, 4, 2, 7, 3766, 5, 723, 36, 71, 43, 530, 476, 26, 400, 317, 46, 7, 4, 12118, 1029, 13, 104, 88, 4, 381, 15, 297, 98, 32, 2071, 56, 26, 141, 6, 194, 7486, 18, 4, 226, 22, 21, 134, 476, 26, 480, 5, 144, 30, 5535, 18, 51, 36, 28, 224, 92, 25, 104, 4, 226, 65, 16, 38, 1334, 88, 12, 16, 283, 5, 16, 4472, 113, 103, 32, 15, 16, 5345, 19, 178, 32]),
       list([1, 194, 1153, 194, 82

In [ ]:
x_train[0]

In [8]:
y_train

array([1, 0, 0, ..., 0, 1, 0])

### Normalize sequencies

In [9]:
len(x_train[0]) # 218 Words

218

In [10]:
len(x_train[1]) # 189 Words

189

All texts must be lenght = 10

In [11]:
x_train = tf.keras.preprocessing.sequence.pad_sequences(x_train, maxlen=max_len)

In [12]:
len(x_train[0])

100

In [13]:
len(x_train[1])

100

In [14]:
x_test = tf.keras.preprocessing.sequence.pad_sequences(x_test, maxlen=max_len)

# Construct Recurrent Neural Network

## Define Model

In [15]:
model = tf.keras.Sequential()

### Add Embedding Layer

https://iaexpert.academy/2019/04/12/word-embedding-transformando-palavras-em-numeros/

![](https://miro.medium.com/max/700/1*-ktg0HACNF4sEY39Ie9Eiw.png)

input_dim = 20.000 -> Lines

output_dim = 128 -> Columns

Return  [ 20.000 x 128 ]

Input Shape = 100  Max 100 words per Sentence

In [16]:
x_train.shape

(25000, 100)

In [17]:
x_train.shape[1]

100

In [18]:
model.add(tf.keras.layers.Embedding(input_dim=number_of_words,output_dim=128,input_shape=[x_train.shape[1]]))

### Add LSTM Layer

units / neurons = 128

activate = tanh

![](https://miro.medium.com/max/700/1*ZafDv3VUm60Eh10OeJu1vw.png)

In [19]:
model.add(tf.keras.layers.LSTM(units=128,activation='tanh'))

### Add Output Layer

Sigmoid return Probability

In [20]:
model.add(tf.keras.layers.Dense(units=1,activation='sigmoid'))

## Compile Model
Binary CrossEntropy == 2 Classes

In [21]:
model.compile(optimizer='rmsprop',loss='binary_crossentropy',metrics=['accuracy'])

In [22]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding (Embedding)        (None, 100, 128)          2560000   
_________________________________________________________________
lstm (LSTM)                  (None, 128)               131584    
_________________________________________________________________
dense (Dense)                (None, 1)                 129       
Total params: 2,691,713
Trainable params: 2,691,713
Non-trainable params: 0
_________________________________________________________________


2,691,713 Weights

## Train Model

batch_size = Weights Update each Registers

In [23]:
model.fit(x_train,y_train,epochs=3,batch_size=128)  

Epoch 1/3
196/196 [==============================] - 13s 25ms/step - loss: 0.5600 - accuracy: 0.7178
Epoch 2/3
196/196 [==============================] - 5s 24ms/step - loss: 0.2900 - accuracy: 0.8834
Epoch 3/3
196/196 [==============================] - 5s 23ms/step - loss: 0.2198 - accuracy: 0.9161


## Evaluate Model

In [24]:
test_loss, test_accuracy = model.evaluate(x_test,y_test)

782/782 [==============================] - 4s 4ms/step - loss: 0.3798 - accuracy: 0.8477


In [25]:
print(test_accuracy)

0.8477200269699097


In [26]:
test_loss

0.37982413172721863